### ISSUE TITLE
Check domain use in concept set in individual qualifying criteria - enhancement request

### ISSUE THREAD

**Issue by jhardin29** [OHDSI/Atlas#474]
### Expected behavior
Create a qualifying criteria or individual rule using a concept set that contains multiple domains. If mistakenly use a single domain and not multiple, ATLAS should change the rule to use the domains in the contained in the concept set or raise this issue to the user. 
### Actual behavior
Create a qualifying criteria or individual rule using a concept set that contains multiple domains. If mistakenly use a single domain and not multiple the cohort generates without identifying the missing domains. 
### Steps to reproduce behavior
Create a qualifying criteria or individual rule using a concept set that contains multiple domains. If mistakenly use one domain and not multiple, potentially identifies fewer numbers of subjects. 
Tagging @ericaVoss as this would aid ATLAS users with vocabulary changes.

**Comment by chrisknoll** [comment]
Hi, @jhardin29 ,
  This can't be done. Each domain criteria has it's own set of parameters (ConditionOccurence has ConditionType, Drug Exposure has DaysSupply, RouteConcept, etc).  It's up to the author of the definition to specify which domain in the CDM to look for the information.

-Chris

**Comment by ericaVoss** [comment]
Is it possible to know if you have a concept set that has concepts that have changed domain since the last time you built?  Would this be helpful?

**Comment by rmakadia** [comment]
Thanks Chris, could a alert be put in that alerts the user that if they had a concept set that doesn't match the domain that they are looking in? So for example, the procedure table is being queried but my concept set is all conditions, could a alert pop up telling me to check my concept set? Thanks!

**Comment by chrisknoll** [comment]
I can think about options about giving the user feedback on their concept set.

However, I feel you guys are barking up the wrong tree here.  The main reason this issue is coming to light is that between vocabulary releases, concepts are bouncing between domains in unpredictable ways.  The problems you're seeing in cohort definitions and concept set expressions is a symptom, not the problem.  I'm sure you all have experience with dealing with a certain source data provider where between refreshes they change column names and column definitions.  I'm sure you are also aware of the havoc this produces with our ETL processes, and while it would be ideal if we could write some software to automatically account for the changes in the upstream provider, I think we all agree that the effort to do this would be monumental.

Unlike our access to those unnamed upstream data providers, we actually have access to close personal friends in our CDM and Vocabulary working groups which we can reach out to them and ask them to get a handle on the situation with concepts and domains.  If they could see that changing concepts' domain is akin to changing the structure of our data tables, they might take more care allowing those sorts of changes to leak through their vocabulary build process.  

However, passing that buck down to me and asking me to fix that problem, I don't think is reasonable.  I think it is reasonable to ask for some sort of UI cue about the contents of a concept set that the author can then make the decision about reviewing the concept set, but I don't think we should throw popups in the face of the user telling them they are doing something wrong when we should assume that the author is knowledgeable in the tool and is making reasonable decisions.  So, if we're going to talk about enhancing the UI, I'd like it to be in the spirit of providing information to the user vs. alerting them that they've done something wrong.

**Comment by rmakadia** [comment]
Thank you Chris, we are simply probing you for options and giving some potential ideas about how to alert the user when choosing multiple domains or domains that do not match. There used to be an alert when creating a concept set that you have multiple domains in the concept set however that has since been removed, but something to that affect would actually be a help to those that are looking for it, otherwise users can ignore it if they want. Never did I say we are telling the user that these are errors, rather they are places to think about and are places that potential errors can occur when doing an analyses. 

I agree when doing anything with the data it is the user who is responsible for the inputs and one cannot pass that on to the tool......

**Comment by chrisknoll** [comment]
Hey, @rmakadia , sorry I didn't mean to assume you were saying that you were telling users of the errors, I was just interpreting 'show a popup alerting the user' as that sort of thing.  The idea about giving the user information about the domains in the concept set is a good one and I'm sure we can come up with a way that isn't obtrusive like 'close this alert box' and more of little indicators next to the concept set representing the domains in use.  I think that'd be helpful and a source of 'reassurance' to the user that what they are doing is what they mean to do.

**Comment by chrisknoll** [comment]
No, the issue here is requesting the atlas UI change the rules based on the values in the concept set.  We should have it on record that we're not going to do that, and this issue should be closed.  Separately, issues around UI hints or cues about contents of the concept set can be introduced and we can track the discussion and implementation as a separate issue.

@jhardin29 , if you agree that the issue you raised is closed, then please close this ticket.

**Comment by jhardin29** [comment]
Hi Chris, After I discuss the options on Friday with Frank we can re assess closure of this thread. Tagging @fdefalco

**Comment by jhardin29** [comment]
Redirecting to new thread with more directed issues and associated explanations but will reference this issue.

**Comment by cgreich** [comment]
@chrisknoll:

"between vocabulary releases, concepts are bouncing between domains in unpredictable ways"!!! I love it. :)

It's called "fixing wrong domain assignments". So, they will keep "bouncing". Unless you want conditions to end up in the device table and the other way around. I can explain why it is so easy for the wrong domain to be assigned to begin with (briefly, because we use heuristics, because you cannot do it by hand for 5M concepts, and the heuristics make mistakes). 

Bottom line: Something of an update button or at least a warning that concepts used in a cohort definition might have changed could be a good idea until we get the bouncing under full control. Even after that.

**Comment by chrisknoll** [comment]
@cgreich,
  Thanks for chiming in. We do have a separate issue to discuss how we could show issues with concept sets and what rules we might want to check in #498.  On the note of updating concepts, shouldn't a new domain assignment belong to a new concept ID?  For the same reason that you wouldn't reuse a concept ID that was a device and use it now as a condition, shouldn't you leave the standard concepts as-is and if something new appears (like a measurement of anemia that's distinct from a diagnosis of anemia) it should be identified separately?

**Comment by cgreich** [comment]
@chrisknoll:

You could do it that way. Right now, the concept_id are linked to the concept_code/vocabulary_id combo of the source. Any other change will just change the record. 

If we did it your way, we would drown in deprecated concepts and "concept_replaced_by" relationships, because the content of the concepts do change. Most of the changes are not ours, but theirs. CMS still makes and deprecates quite a lot of ICD-10-CM codes (they are leaving ICD-9-CM alone these days thank God.

But speaking of, that Update button should follow those replacement links as well. But let me go to #498 now.

**Comment by cgreich** [comment]
@jhardin29, @ericaVoss:

I love it. #498 redirects to #474 ,and #474 redirects to the #498. It's called the cat is biting into her tail. Which one are we using to discuss this topic now? :)

**Comment by jhardin29** [comment]
@cgreich we are using #498 but there is one reference you missed #283 :)
